src/store/epics/chatEpics.js

```diff
@@ -76,31 +81,100 @@ const initChatHandlerConnection = ({ authToken, secretKey, mID, isNewRoom }) =>
+      miniLock.session.keyPairReady = true;
+      let mID = miniLock.crypto.getMiniLockID(keyPair.publicKey);
+      // 4. When we have keypair, login:
+
+      if (isNewPassphrase) {
+        document.location.hash = '#' + passphrase;
+      }
+
+      observer.next({ passphrase, keyPair, mID, isNewRoom: isNewPassphrase });
+      observer.complete();
+    });
+  });
+}
+
+function createDecryptMessageObservable({ message, mID, secretKey }) {
+  return Observable.create(function (observer) {
+    miniLock.crypto.decryptFile(message, mID, secretKey,
+      function (fileBlob, saveName, senderID) {
+        const reader = new FileReader();
+        reader.addEventListener("loadend", () => {
+          const authToken = reader.result;
+          observer.next(authToken);
+          observer.complete();
+        });
+
+        reader.readAsText(fileBlob);
+      });
+  });
+}
+
+function getAuthRequestSettings({ mID }) {
+  const settings = {
+    url: authUrl,
+    responseType: 'blob',
+    headers: {
+      'X-Minilock-Id': mID
+    },
+    method: 'GET'
+  };
+  return settings;
+}
+
 const connectionAlertTtlSeconds = 4;
 
 const initConnectionEpic = (action$) =>
   action$.ofType(CHAT_INIT_CONNECTION)
-    .mergeMap(initChatHandlerConnection)
-    .mergeMap((isNewRoom) =>
-      Observable.merge(
-        chatHandler.getMessageSubject()
-          .map(addMessage),
-
-        chatHandler.getUserStatusSubject()
-          .map(setUserStatus),
-
-        Observable.of(
-          connectionInitiated(),
-          alertSuccess(`${isNewRoom ? 'New room created.' : ''} Connected to server.`, connectionAlertTtlSeconds)
+    .mergeMap(createKeyPairObservable)
+    .mergeMap(({ keyPair, mID, isNewRoom, passphrase }) =>
+      ajax(getAuthRequestSettings({ mID }))
+        .map(ajaxResult => ({
+          message: ajaxResult.response,
+          mID,
+          secretKey: keyPair.secretKey
+        }))
+        .mergeMap(createDecryptMessageObservable)
+        .mergeMap(authToken =>
+          chatHandler.initConnection({
+            authToken,
+            secretKey: keyPair.secretKey,
+            mID
+          })
         )
-      )
-    ).catch(error => {
-      console.error(error);
-      return Observable.from([
-        alertWarning('Lost connection to chat server. Trying to reconnect...'),
-        disconnected()
-      ]);
-    })
+        .mergeMap(() =>
+          Observable.merge(
+            chatHandler.getMessageSubject()
+              .map(addMessage),
+
+            chatHandler.getUserStatusSubject()
+              .map(setUserStatus),
+
+            Observable.of(connectionInitiated(), alertSuccess(`${isNewRoom ? 'New room created.' : ''} Connected to server.`, connectionAlertTtlSeconds))
+          )
+        ).catch(error => {
+          console.error(error);
+          return Observable.from([
+            alertWarning('Lost connection to chat server. Trying to reconnect...'),
+            disconnected()
```

Comment:
This sets the app to `shouldConnect`, which prompts a reconnection attempt, in part. This catch block is what initiates the retry, if the ajax request fails. This is what needs to be re-created, functionally, over in `sessions.js`.

Reply:
Gotcha